File: plugin/src/main/java/co/uk/magmo/puretickets/ticket/TicketManager.java
package co.uk.magmo.puretickets.ticket;

import co.uk.magmo.puretickets.configuration.Config;
import co.uk.magmo.puretickets.exceptions.TicketClosed;
import co.uk.magmo.puretickets.exceptions.TicketOpen;
import co.uk.magmo.puretickets.exceptions.TooManyOpenTickets;
import co.uk.magmo.puretickets.storage.SQLManager;
import co.uk.magmo.puretickets.storage.TimeAmount;
import com.google.common.collect.Lists;
import org.bukkit.Location;
import org.bukkit.entity.Player;
import org.jetbrains.annotations.Nullable;

import java.time.LocalDateTime;
import java.util.*;

public class TicketManager {
    private final Config config;
    private final SQLManager sqlManager;

    public TicketManager(Config config, SQLManager sqlManager) {
        this.config = config;
        this.sqlManager = sqlManager;
    }

    public @Nullable Ticket get(int id) {
        return sqlManager.getTicket().select(id);
    }

    public List<Ticket> getAll(UUID uuid, TicketStatus status) {
        return sqlManager.getTicket().selectAll(uuid, status);
    }

    public Ticket getLatestTicket(UUID uuid, TicketStatus... statuses) {
        return sqlManager.getTicket().selectLastTicket(uuid, statuses);
    }

    public List<Integer> getIds(UUID uuid, TicketStatus status) {
        return sqlManager.getTicket().selectIds(uuid, status);
    }

    public Boolean exists(Integer id) {
        return sqlManager.getTicket().exists(id);
    }

    public Integer count(UUID uuid) {
        return sqlManager.getTicket().count(uuid);
    }

    public Integer count(TicketStatus status) {
        return sqlManager.getTicket().count(status);
    }

    public EnumMap<TicketStatus, Integer> stats(UUID uuid) {
        return sqlManager.getTicket().selectTicketStats(uuid);
    }

    public List<Ticket> all(TicketStatus status) {
        return sqlManager.getTicket().selectAll(status);
    }

    public List<String> allNames(TicketStatus status) {
        return sqlManager.getTicket().selectNames(status);
    }

    public HashMap<UUID, Integer> highscores(TimeAmount span) {
        return sqlManager.getTicket().highscores(span);
    }

    public Ticket createTicket(Player player, Message message) throws TooManyOpenTickets {
        if (count(player.getUniqueId()) >= config.LIMIT__OPEN_TICKETS) {
            throw new TooManyOpenTickets(config);
        }

        UUID uuid = player.getUniqueId();
        Location location = player.getLocation();

        Integer id = sqlManager.getTicket().insert(uuid, TicketStatus.OPEN, null, location);
        Ticket ticket = new Ticket(id, uuid, Lists.newArrayList(message), location, TicketStatus.OPEN, null);

        sqlManager.getMessage().insert(ticket, message);

        return ticket;
    }

    public Ticket update(Ticket ticket, Message message) throws TicketClosed {
        if (ticket.getStatus() == TicketStatus.CLOSED) {
            throw new TicketClosed();
        }

        return addMessageAndUpdate(ticket, message);
    }

    public Ticket pick(UUID uuid, Ticket ticket) throws TicketClosed {
        if (ticket.getStatus() == TicketStatus.CLOSED) {
            throw new TicketClosed();
        }

        Message message = new Message(MessageReason.PICKED, LocalDateTime.now(), uuid);

        ticket.setStatus(TicketStatus.PICKED);
        ticket.setPickerUUID(uuid);

        return addMessageAndUpdate(ticket, message);
    }

    public Ticket yield(UUID uuid, Ticket ticket) throws TicketOpen {
        if (ticket.getStatus() == TicketStatus.OPEN) {
            throw new TicketOpen();
        }

        Message message = new Message(MessageReason.REOPENED, LocalDateTime.now(), uuid);

        ticket.setStatus(TicketStatus.OPEN);
        ticket.setPickerUUID(uuid);

        return addMessageAndUpdate(ticket, message);
    }

    public Ticket close(UUID uuid, Ticket ticket) throws TicketClosed {
        if (ticket.getStatus() == TicketStatus.CLOSED) {
            throw new TicketClosed();
        }

        Message message = new Message(MessageReason.CLOSED, LocalDateTime.now(), uuid);

        ticket.setStatus(TicketStatus.CLOSED);

        return addMessageAndUpdate(ticket, message);
    }

    public Ticket done(UUID uuid, Ticket ticket) throws TicketClosed {
        if (ticket.getStatus() == TicketStatus.CLOSED) {
            throw new TicketClosed();
        }

        Message message = new Message(MessageReason.DONE_MARKED, LocalDateTime.now(), uuid);

        ticket.setStatus(TicketStatus.CLOSED);

        return addMessageAndUpdate(ticket, message);
    }

    public Ticket reopen(UUID uuid, Ticket ticket) throws TicketOpen {
        if (ticket.getStatus() == TicketStatus.OPEN) {
            throw new TicketOpen();
        }

        Message message = new Message(MessageReason.REOPENED, LocalDateTime.now(), uuid);

        ticket.setStatus(TicketStatus.OPEN);

        return addMessageAndUpdate(ticket, message);
    }

    public Ticket note(UUID uuid, Ticket ticket, String input) {
        Message message = new Message(MessageReason.NOTE, LocalDateTime.now(), input, uuid);

        return addMessageAndUpdate(ticket, message);
    }

    private Ticket addMessageAndUpdate(Ticket ticket, Message message) {
        ticket.getMessages().add(message);

        sqlManager.getMessage().insert(ticket, message);
        sqlManager.getTicket().update(ticket);

        return ticket;
    }
}

File: src/main/kotlin/co/uk/magmo/puretickets/ticket/TicketManager.kt
package co.uk.magmo.puretickets.ticket

import co.uk.magmo.puretickets.storage.SQLManager
import org.bukkit.OfflinePlayer
import org.bukkit.entity.Player
import java.util.*

class TicketManager(private val sqlManager: SQLManager) {
    private val tickets = sqlManager.ticket.selectActive()
    private var current = sqlManager.ticket.currentId()

    operator fun get(uuid: UUID?, id: Int): Ticket? = tickets[uuid].first { ticket -> ticket.id == id }

    operator fun get(uuid: UUID?): List<Ticket> = tickets[uuid]

    operator fun contains(player: OfflinePlayer) = tickets.containsKey(player.uniqueId)

    fun asMap(): MutableMap<UUID, MutableCollection<Ticket>> = tickets.asMap()

    fun allKeys(): MutableSet<UUID> = tickets.keySet()

    fun createTicket(player: Player, message: Message): Ticket {
        current += 1
        return Ticket(current, player.uniqueId, arrayListOf(message), TicketStatus.OPEN, null, player.location).also {
            tickets.put(player.uniqueId, it)

            sqlManager.ticket.insert(it)
            sqlManager.message.insert(it, message)
        }
    }

    fun update(information: TicketInformation, message: Message): Ticket {
        val ticket = tickets[information.player].first { ticket -> ticket.id == information.index }
        val index = tickets[information.player].indexOf(ticket)

        ticket.addMessageAndUpdate(message)
        tickets[information.player][index] = ticket

        return ticket
    }

    fun pick(uuid: UUID?, information: TicketInformation): Ticket {
        val ticket = tickets[information.player].first { ticket -> ticket.id == information.index }
        val index = tickets[information.player].indexOf(ticket)
        val message = Message(MessageReason.PICKED, null, uuid)

        ticket.status = TicketStatus.PICKED
        ticket.pickerUUID = uuid
        ticket.addMessageAndUpdate(message)

        tickets[information.player][index] = ticket
        return ticket
    }

    fun yield(uuid: UUID?, information: TicketInformation): Ticket {
        val ticket = tickets[information.player].first { ticket -> ticket.id == information.index }
        val index = tickets[information.player].indexOf(ticket)
        val message = Message(MessageReason.YIELDED, null, uuid)

        ticket.status = TicketStatus.OPEN
        ticket.pickerUUID = null
        ticket.addMessageAndUpdate(message)

        tickets[information.player][index] = ticket

        return ticket
    }

    fun close(uuid: UUID?, information: TicketInformation): Ticket {
        val ticket = tickets[information.player].first { ticket -> ticket.id == information.index }
        val message = Message(MessageReason.CLOSED, null, uuid)

        tickets.remove(information.player, ticket)
        ticket.status = TicketStatus.CLOSED
        ticket.addMessageAndUpdate(message)

        return ticket
    }

    fun done(uuid: UUID?, information: TicketInformation): Ticket {
        val ticket = tickets[information.player].first { ticket -> ticket.id == information.index }
        val message = Message(MessageReason.DONE_MARKED, null, uuid)

        tickets.remove(information.player, ticket)
        ticket.status = TicketStatus.CLOSED
        ticket.addMessageAndUpdate(message)

        return ticket
    }

    fun reopen(uuid: UUID?, information: TicketInformation): Ticket {
        val ticket = sqlManager.ticket.select(information.index)
        val message = Message(MessageReason.REOPENED, null, uuid)

        ticket.status = TicketStatus.OPEN
        ticket.addMessageAndUpdate(message)

        tickets.put(information.player, ticket)

        return ticket
    }

    fun note(uuid: UUID?, information: TicketInformation, input: String): Ticket {
        val ticket = tickets[information.player].first { ticket -> ticket.id == information.index }
        val message = Message(MessageReason.NOTE, input, uuid)

        ticket.addMessageAndUpdate(message)

        return ticket
    }

    private fun Ticket.addMessageAndUpdate(message: Message) {
        messages.add(message)

        sqlManager.message.insert(this, message)
        sqlManager.ticket.update(this)
    }
}